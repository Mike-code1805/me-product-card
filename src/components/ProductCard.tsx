import { useProduct } from '../hooks/useProduct';
import styles from '../styles/styles.module.css';
import React, { createContext, CSSProperties } from 'react';
import {
  InitialValues,
  onChangeArgs,
  ProductCartHandlers,
} from '../interfaces/products.interfaces';
import {
  Product,
  ProductContextProps,
} from '../interfaces/products.interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children?: (args: ProductCartHandlers) => JSX.Element | null;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    product,
    onChange,
    value,
    initialValues,
  });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        maxCount,
        product,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children!({
          count: counter,
          isMaxCountReached: isMaxCountReached,
          maxCount,
          product,
          reset,
          increaseBy,
        })}
      </div>
    </Provider>
  );
};
