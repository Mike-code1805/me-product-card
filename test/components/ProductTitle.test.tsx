import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle } from '../../src/components/ProductTitle';
import { product1 } from '../data/product';
import { ProductCard } from '../../src/components/index';

describe('ProductTitle', () => {
  test('should render product title component', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Custom Product Title" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should show component with product title name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
