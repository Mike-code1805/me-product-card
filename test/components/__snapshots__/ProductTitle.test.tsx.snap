// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should render product title component 1`] = `
<span
  className="productDescription undefined"
>
  Custom Product Title
</span>
`;

exports[`ProductTitle should show component with product title name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Card
  </span>
</div>
`;
